### DISCUSSION TITLE
Some ideas about attribute

### DISCUSSION THREAD

**Discussion by Bsntmnhss** [dotnet/csharplang#4832] [Language Ideas]
## Add a new keyword "attribute" to create an attribute
For example:
``` csharp
[AttributeUsage(...)]
public attribute MyAttribute
{
    ...
}
```
It will translate into:
~~~ csharp
[AttributeUsage(...)]
public class MyAttribute : Attribute
{
    ...
}
~~~
## Add more attribute targets 
Such as "Record" "Attribute" and so on 
``` csharp
[Flags]
public enum AttributeTargets
{
    ...
    Record,
    Attribute,
    ...
}
```
Now the AttributeUsage attribute can become like this:
~~~ csharp
[AttributeUsage(AttributeTargets.Attribute, Inherited = true)]
public attribute AttributeUsageAttribute
{
    ...
}
~~~
## Add some attribute event methods
For example:
``` csharp
[AttributeUsage(AttributeTargets.Method)]
public attribute TestMethodAttribute
{
    public override void BeforeInvokeMethod(MethodInfo info, object? sender, object?[]? args ...)
    {
        ...
    }
}


public class Program
{
    [TestMethod]
    static void Test()
    {
        ...
    }
    
    static void Main(string[] args)
    {
        Test(); //Call the "BeforeInvokeMethod" method then call the "Test" method
    }
}
```
This is an example of "BeforeInvokeMethod" event method
Others (for example, fields and properties have OnGet, OnSet and other event methods)

**Comment by YairHalberstadt** [comment]
What is the motivation behind these ideas? What is the real world use case where current solutions are insufficient, and these suggestions will help?

**Comment by huoyaoyuan** [comment]
> Add a new keyword "attribute" to create an attribute

Usable, but doesn't solve a lot comparing to current syntax.

> Add more attribute targets

> Such as "Record"

This is something we intentionally avoided. The team said it should be possible to use non-record code to define a class same to record.

> "Attribute"

This is really only meaningful for `AttributeUsageAttribute`. Since we can't change that attribute, it will be useless at all.

> Add some attribute event methods

This are better for a custom code generator recognizing the attribute.
